[Plugin|plugins] Consolidate command collection API methods

TurboTurtle opened the pull request:
Removed `call_ext_prog()`, `check_ext_prog()`, and
`get_command_output()` in favor of consolidating Plugin command
collection into two public methods: `add_cmd_output()` and
`get_cmd_output_now()`.

- `add_cmd_output()` remains unchanged and is the interface to use when
plugins should collect command output during their collect() phase.

- `get_cmd_output_now()` will now return the content of a command's
execution as well as the filename within the sos archive, instead of
just returning the filename. This should help streamline plugin code
that runs a command and then iterates over that output.

The `get_cmd_output_now()` has a `save` keyword arg (default True), that
if set to False will skip writing the command output to the archive.

Also updates the plugins to no longer use `get_command_output()`,
`call_ext_prog()`, or `check_ext_prog()` - instead only using
`get_cmd_output_now()`.

For the vast majority of plugins, these calls were able to be changed
in-place with little to no further alteration of plugin code outside of
no longer making indenpendent open() calls to files created, as the
content of those files is now directly available. Exceptions are noted
below.

For the dlm, gfs2, and kvm plugins the only use of get_command_output()
was to mount /sys/kernel/debugfs which in modern distributions is no
longer needed, so those plugins have had that functionality entirely
removed instead of being ported to the new `get_cmd_output_now()`.

For the networking plugin, many of the checks are handled by
SoSPredicates, and so those command collections now use Predicates
rather than being gated by individual `get_cmd_output_now()` calls.


---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?

TurboTurtle commented:
This is a continuation from the conversation in #1707.

Also, it looks like this should address #1754 as a bonus, as `add_cmd_output()` eventually calls `_get_cmd_output_now()` which is where the docstring has been expanded.

pmoravec commented on sos/plugins/abrt.py:
```diff
@@ -24,16 +24,15 @@ class Abrt(Plugin, RedHatPlugin):
-        if self.get_option('detailed') and list_file:
-            self.info_detailed(list_file)
+        abrt_list = self.get_cmd_output_now("abrt-cli list")
+        if self.get_option("detailed") and abrt_list:
```
Technically, abrt_list (as a return value of `get_cmd_output_now`) will always be a dict, so `and abrt_list` will always pass. We have few options here:

- test for `abrt_list["status"] == 0`
- test for `abrt_list["output"]` - that is bit redundant due to the further iteration over output splitlines
- skip this test

I am inclined to skip the test.

TurboTurtle replied:
Yep, oversight on my part - should be `abrt_list['status'] == 0` here.

pmoravec commented on sos/plugins/block.py:
```diff
@@ -69,10 +53,12 @@ def setup(self):
-        if lsblk_file:
-            for dev in self.get_luks_devices(lsblk_file):
-                self.add_cmd_output('cryptsetup luksDump /dev/%s' % dev)
+        if lsblk['status'] == 0:
```
See my comment in `abrt` - ideally, we shall be coherent here (no strong prferences, either testing for zero status or skipping such a test is fine to me).

TurboTurtle replied:
I'm definitely in the "always check status == 0" camp.

We have no other way to gate against the iterations that follow from executing against error output when the command fails for one reason or another.

pmoravec commented on sos/plugins/crio.py:
```diff
@@ -84,7 +84,7 @@ def setup(self):
         ret = []
-        result = self.get_command_output(cmd)
+        result = self.get_cmd_output_now(cmd, save=False)
         if result['status'] == 0:
```
ditto here (test status or skip the test)

pmoravec commented on sos/plugins/dlm.py:
```diff
@@ -37,36 +33,17 @@ def setup(self):
-        return
+        dlm_tool = "dlm_tool ls"
+        result = self.get_cmd_output_now(dlm_tool)
+        if result["status"] != 0:
```
ditto here (test status or skip the test)

pmoravec commented on sos/plugins/dnf.py:
```diff
@@ -57,11 +57,11 @@ def setup(self):
-            history = self.call_ext_prog("dnf history")
+            history = self.get_cmd_output_now("dnf history")
             transactions = -1
             if history['output']:
```
Although it is not directly relevant to the change in API, could we drop this test? IMHO iterating over (potentially empty) splitlines is sufficient..

TurboTurtle replied:
Fair point, will update

pmoravec commented on sos/plugins/docker.py:
```diff
@@ -104,7 +103,7 @@ def setup(self):
         ret = []
-        result = self.get_command_output(cmd)
+        result = self.get_cmd_output_now(cmd, save=False)
         if result['status'] == 0:
```
ditto here (test for status or no test)

pmoravec commented on sos/plugins/gfs2.py:
```diff
@@ -36,24 +32,6 @@ def setup(self):
         ])
 
         if self.get_option("gfs2lockdump"):
-            if self._mount_debug():
```
Wasn't there some rationale to collect `/sys/kernel/debug/gfs2/*` _only_ under some further condition (i.e. `_mount_debug`)?

pmoravec commented on sos/plugins/gfs2.py:
```diff
@@ -36,24 +32,6 @@ def setup(self):
-
-    def postproc(self):
-        if self._debugfs_cleanup and os.path.ismount(self.debugfs_path):
-            r = self.call_ext_prog("umount %s" % self.debugfs_path)
```
Is it really ok not to umount (something we probably mounted)?

pmoravec commented on sos/plugins/gluster.py:
```diff
@@ -108,14 +94,17 @@ def setup(self):
                                  "directory: %s" % self.statedump_dir)
-            state = self.get_command_output("gluster get-state")
+            state = self.get_cmd_output_now("gluster get-state")
             if state['status'] == 0:
```
ditto here

pmoravec commented on sos/plugins/gluster.py:
```diff
@@ -108,14 +94,17 @@ def setup(self):
-        if volume_file:
-            for volname in self.get_volume_names(volume_file):
+        volume_cmd = self.get_cmd_output_now("gluster volume info")
+        if volume_cmd['status'] == 0:
```
ditto here

pmoravec commented on sos/plugins/kata_containers.py:
```diff
@@ -37,7 +37,7 @@ def setup(self):
         cmd = 'kata-runtime --kata-show-default-config-paths'
-        configs = self.get_command_output(cmd)
+        configs = self.get_cmd_output_now(cmd)
         if configs and configs['status']:
```
"`configs`" is never evaluated to `False`.

pmoravec commented on sos/plugins/kvm.py:
```diff
@@ -33,19 +30,7 @@ def setup(self):
             "/sys/module/kvm_amd/srcversion",
             "/sys/module/ksm/srcversion"
         ])
-        if not os.path.ismount(self.debugfs_path):
```
Same comments like to `gfs2` plugin - is it OK in skipping this test?

pmoravec commented on sos/plugins/openstack_glance.py:
```diff
@@ -76,7 +72,10 @@ def setup(self):
     def running_in_container(self):
         for runtime in ["docker", "podman"]:
-            container_status = self.get_command_output(runtime + " ps")
+            container_status = self.get_cmd_output(
```
to call `self.get_cmd_output_now` instead? :)

TurboTurtle replied:
Typo, yep - thanks for catching that.

pmoravec commented on sos/plugins/pcp.py:
```diff
@@ -148,10 +148,8 @@ def setup(self):
-        self.add_cmd_output("pcp")
         # Collect a summary for the current day
-        res = self.get_command_output('pcp')
+        res = self.get_cmd_output_now('pcp')
```
with `save=False`?

TurboTurtle replied:
We previously ran `add_cmd_output('pcp')`, which would be redundant with the new approach of `get_cmd_output_now()`. So  `save=False` would be a regression here.

pmoravec commented on sos/plugins/sas3ircu.py:
```diff
@@ -23,11 +23,9 @@ class SAS3ircu(Plugin, RedHatPlugin, DebianPlugin, UbuntuPlugin):
 
         # get list of adapters
-        result = self.call_ext_prog("sas3ircu list", timeout=5)
+        result = self.get_cmd_output_now("sas3ircu list", timeout=5)
```
with `save=False`?

pmoravec commented on sos/plugins/selinux.py:
```diff
@@ -27,7 +27,7 @@ def setup(self):
         self.add_cmd_output('sestatus')
 
-        state = self.get_command_output('getenforce')['output']
+        state = self.get_cmd_output_now('getenforce')['output']
```
We will newly collect/store output of this command - a duplicate info resent in other cmd output(s), but it simplifies to find the info. Not sure if worth having `save=False` or not.

pmoravec commented on sos/plugins/teamd.py:
```diff
@@ -21,7 +21,7 @@ class Teamd(Plugin, RedHatPlugin, DebianPlugin, UbuntuPlugin):
     def _get_team_interfaces(self):
         teams = []
-        ip_result = self.get_command_output("ip -o link")
+        ip_result = self.get_cmd_output_now("ip -o link")
```
with `save=False`?

pmoravec commented on sos/plugins/vdsm.py:
```diff
@@ -114,7 +114,7 @@ def setup(self):
 
         try:
-            res = self.call_ext_prog(
+            res = self.get_cmd_output_now(
```
with `save=False`?

TurboTurtle replied:
Collecting the connected storage pools output would be beneficial to RH support teams

pmoravec commented on sos/plugins/vdsm.py:
```diff
@@ -130,7 +130,7 @@ def setup(self):
 
         try:
-            res = self.call_ext_prog('vdsm-client Host getStorageDomains')
+            res = self.get_cmd_output_now('vdsm-client Host getStorageDomains')
```
with `save=False`?

TurboTurtle replied:
Same here, this is useful data for support teams

pmoravec commented on sos/plugins/vhostmd.py:
```diff
@@ -20,7 +20,7 @@ class vhostmd(Plugin, RedHatPlugin):
 
     def setup(self):
-        vw = self.get_command_output("virt-what")['output'].splitlines()
+        vw = self.get_cmd_output_now("virt-what")['output'].splitlines()
```
with `save=False`?

pmoravec commented on sos/plugins/virsh.py:
```diff
@@ -43,8 +43,7 @@ def setup(self):
         for k in ['net', 'nwfilter', 'pool']:
-            self.add_cmd_output('%s %s-list' % (cmd, k))
-            k_list = self.get_command_output('%s %s-list' % (cmd, k))
+            k_list = self.get_cmd_output_now('%s %s-list' % (cmd, k))
```
with `save=False`?

TurboTurtle replied:
Same as `pcp` - without `save=False` we "automatically" get the `add_cmd_output()` call from above.

pmoravec commented on sos/plugins/virsh.py:
```diff
@@ -55,7 +54,7 @@ def setup(self):
         # cycle through the VMs/domains list, ignore 2 header lines and latest
         # empty line, and dumpxml domain name in 2nd column
-        domains_output = self.get_command_output('%s list --all' % cmd)
+        domains_output = self.get_cmd_output_now('%s list --all' % cmd)
```
with `save=False`?

TurboTurtle replied:
Same here

pmoravec commented on sos/plugins/xen.py:
```diff
@@ -37,7 +42,7 @@ def check_enabled(self):
 
     def is_running_xenstored(self):
-        xs_pid = self.call_ext_prog("pidof xenstored")['output']
+        xs_pid = self.get_cmd_output_now("pidof xenstored")['output']
```
with `save=False`?

TurboTurtle replied:
Ack

pmoravec commented on sos/plugins/yum.py:
```diff
@@ -99,7 +99,7 @@ def setup(self):
             # self.add_cmd_output("yum-debug-dump '%s'"
             # % os.path.join(self.commons['dstroot'],"yum-debug-dump"))
-            r = self.call_ext_prog("yum-debug-dump")
+            r = self.get_cmd_output_now("yum-debug-dump")
```
with `save=False`?

TurboTurtle replied:
Ack

pmoravec commented on sos/plugins/zfs.py:
```diff
@@ -27,7 +27,7 @@ def setup(self):
         ])
 
-        zpools = self.call_ext_prog("zpool list -H -o name")
+        zpools = self.get_cmd_output_now("zpool list -H -o name")
```
with `save=False`? (though this might be interesting to have..? (dont know zfs..))

pmoravec commented on sos/plugins/__init__.py:
```diff
@@ -1112,59 +1061,123 @@ def _get_cmd_output_now(self, cmd, suggest_filename=None,
+
+        if result['status'] == 124:
+            self._log_warn(
+                "command '%s' timed out after %ds" % (prog, timeout)
```
copy&paste typo:  s/prog/cmd

pmoravec commented:
In general, I like the consolidation. Good work.

We are not coherent in calling `if something["status"] == 0:` - usually such a test is redundant (but dont harm), but in few cases it is afaik really needed. Not sure if it makes sense to unify the approach here.

Several other plugins needs to be updated as well - see their backtraces (usually due to different return value of `add_cmd_output` (not file name but dict)):

    composer-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/composer.py", line 30, in setup
        blueprints = self._get_entries("composer-cli blueprints list")
      File "/root/sos-master-1807/sos/plugins/composer.py", line 17, in _get_entries
        with open(ent_file, "r") as ents:
    TypeError: coercing to Unicode: need string or buffer, dict found
    
    kpatch-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/kpatch.py", line 27, in setup
        kpatches = open(kpatch_list, "r").read().splitlines()
    TypeError: coercing to Unicode: need string or buffer, dict found
    
    openstack_keystone-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/openstack_keystone.py", line 117, in setup
        super(RedHatKeystone, self).setup()
      File "/root/sos-master-1807/sos/plugins/openstack_keystone.py", line 55, in setup
        not(os.path.isdir(self.domain_config_dir)):
      File "/usr/lib64/python2.7/genericpath.py", line 41, in isdir
        st = os.stat(s)
    TypeError: coercing to Unicode: need string or buffer, dict found
    
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 1167, in postproc
        plug.postproc()
      File "/root/sos-master-1807/sos/plugins/openstack_keystone.py", line 100, in postproc
        r"((?m)^\s*(%s)\s*=\s*)(.*)", r"\1********")
      File "/root/sos-master-1807/sos/plugins/__init__.py", line 595, in do_path_regex_sub
        pathexp = re.compile(pathexp)
      File "/usr/lib64/python2.7/re.py", line 190, in compile
        return _compile(pattern, flags)
      File "/usr/lib64/python2.7/re.py", line 229, in _compile
        p = _cache.get(cachekey)
    TypeError: unhashable type: 'dict'
    
    pcp-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/pcp.py", line 132, in setup
        for line in open(pmlogger_ls).read().splitlines():
    TypeError: coercing to Unicode: need string or buffer, dict found
    
    sapnw-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/sapnw.py", line 134, in setup
        self.collect_list_instances()
      File "/root/sos-master-1807/sos/plugins/sapnw.py", line 40, in collect_list_instances
        p = open(inst_out, "r").read().splitlines()
    TypeError: coercing to Unicode: need string or buffer, dict found

TurboTurtle commented:
After talking with Bryn a bit further, I've updated this PR to also include a rename of the `get_cmd_output_now()` methods to `collect_cmd_output()`, which `add_cmd_output()` still leverages.

Also, dropped the `save` kwarg in favor of a replacement `exec_cmd()` method to replace the `*_ext_prog()` methods.


Looking at the LGTM analysis, I'm not sure where the build failure is coming from.
